### DISCUSSION TITLE
How to add newly added documents incrementally to Milvus.

### DISCUSSION THREAD

**Discussion by gaye746560359** [milvus-io/milvus#24426] [Q&A and General discussion]
I have a directory on my local machine with ten documents, which I have already added to Milvus. Now, if I add a new document to the directory, how can I add only the new document to Milvus? Also, if there is an existing document with the same name but different content, how can I update it in Milvus?

**Comment by yhmo** [comment]
In milvus, each document(we call it entity) has a unique ID, and each entity has only one vector.
Some users wish to prevent ID duplication, some users wish to prevent vector duplication.
- For ID duplication, you can use delete()+insert() to update an exist entity
Assume you have a collection, the primary key field name is "id", now you want to insert a new entity with id=5
1. collection.delete(expr="id==5")
2. collection.insert(new entity)

- For vector duplication, you can call search() before delete()+insert(). If an entity has zero distance from the new entity, delete the old entity and insert the new one.
Assume you have a collection, now you want to insert a new entity. Assume the index is IVF index with nlist=1024, you need to call full search(nprobe=1024) to scan the entire collection, if there is an entity has zero distance from the new entity, delete it and insert the new one
1. collection.search(nprobe=1024, topk=1)
2. if the top1 entity has zero distance from the new entity, the top1 id is 8, call delete(expr="id==8") to delete it
3. collection.insert(new entity)

**Reply by ArijitSinghEDA** [reply]
Is deleting and inserting the only option? What if I want to append to my existing collection?

**Reply by gaye746560359** [reply]
> 在 milvus 中，每个文档（我们称之为实体）都有一个唯一的 ID，每个实体只有一个向量。 有些用户希望防止 ID 重复，有些用户希望防止向量重复。
> 
> * 对于ID重复，可以使用delete()+insert()来更新一个已经存在的实体
>   假设你有一个集合，主键字段名为“id”，现在你想插入一个id=5的新实体
> 
> 1. collection.delete(expr="id==5")
> 2. collection.insert（新实体）
> 
> * 对于向量复制，可以在delete()+insert()之前调用search()。如果实体与新实体的距离为零，则删除旧实体并插入新实体。
>   假设您有一个集合，现在您想插入一个新实体。假设索引为nlist=1024的IVF索引，需要调用full search(nprobe=1024)扫描整个集合，如果有一个实体与新实体的距离为零，则将其删除并插入新实体
> 
> 1. collection.search(nprobe=1024, topk=1)
> 2. 如果 top1 实体与新实体的距离为零，则 top1 id 为 8，调用 delete(expr="id==8") 将其删除
> 3. collection.insert（新实体）

In langchain, how to use the Milvus module? The following is my current code for adding data, which obviously has problems when there are new documents:
from langchain.vectorstores import Milvus
document = DirectoryLoader('../', glob='docs/*.txt', loader_cls=TextLoader,
                           loader_kwargs={"encoding": "utf-8"}).load()
splitter = CharacterTextSplitter(chunk_size=500, chunk_overlap=0).split_documents(document)
Milvus.from_documents(
    splitter,
    embeddings,
    connection_args={"host": "127.0.0.1", "port": "19530"},
)

**Reply by xiaofan-luan** [reply]
> ng the only option? What if I want to append to my existing collection?

I would suggest the id of milvus is chunkID. and docID is another nonPK field.

Does that meet your requirment?

Or do you want to do groupby operation on a document?

**Reply by ArijitSinghEDA** [reply]
> Does that meet your requirment?

No. I was asking if I want to add new data to an existing collection without deleting the whole collection and then inserting it again with the new data, is there any way to do it?

**Reply by xiaofan-luan** [reply]
> > Does that meet your requirment?
> 
> No. I was asking if I want to add new data to an existing collection without deleting the whole collection and then inserting it again with the new data, is there any way to do it?

just do insert should do

**Reply by ArijitSinghEDA** [reply]
> just do insert should do

Yes, it works. I was under the assumption that it clears data every time we do insertion in an existing collection. Thank you.
